Additional Telemetry - Implementation of RFC0036

JamesWTruher opened the pull request:
this is the implementation of https://github.com/PowerShell/PowerShell-RFC/blob/249e8d88eb779a6003fd138609ae94417ae13698/2-Draft-Accepted/RFC0036-AdditionalTelemetry.md 

# PR Summary

* Add Telemetry in the following areas
  * When execution occurs, the ApplicationType will be captured as a metric
  * When a module is loaded, capture a metric of the module name if it's in the allow list.
  * When `PowerShell.Create` is used, send a metric
  * When an experimental feature is activated send a metric (this is at startup only)
  * Update startup telemetry to measure the type of shell startup (remote/ssh)
* Added a couple of new fields to the StartUp telemetry
  * `Channel` allows us to understand when PS is being used in a docker instance and what the OS is.
  * Added a unique identifier which allows us to determine unique users. It's a generated guid which is persisted to disk.
  * Added a unique identifier for the shell session

I added an internal member to `System.Management.Automation.Platform` called `CacheDirectory` which makes it easy to find the location of the cache directory. I think this should probably be public, but that can be decided at a later time.

## PR Context

We knew that our initial telemetry was just that. The bare minimum of information to determine whether the shell was actually being used and on what platform. The additional information we think we need is captured in the RFC.


## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

JamesWTruher committed:
initial pass at new telemetry

JamesWTruher committed:
remove some extraneous spaces at line endings

JamesWTruher committed:
Add more detailed startup data for console startup

change initialization of telemetry client
change the name of the uid generator

JamesWTruher committed:
Fix up name of OSDetail channel

JamesWTruher committed:
Remove pipeline telemetry, change the way the mutex is handled

JamesWTruher committed:
First round of clean up

JamesWTruher committed:
additional refactoring

move most of the property initialization into the static constructor which won't be called in the case that telemetry is not enabled.

JamesWTruher committed:
attempt to send telemetry if there is a catastrophic telemetry failure

Additional cleanup

JamesWTruher committed:
further clean up

JamesWTruher committed:
Add tests for telemetry persistence and correct reading of the environment variable

JamesWTruher committed:
Add telemetry for experimental features

Fix test to find the cache directory on Windows
update a comment for experimental features

JamesWTruher committed:
turn telemetry back off by default for test runs

JamesWTruher committed:
Change public types to internal

allow only CanSendTelemetry to publically visible, but make the setter private so it can't be altered after start.
Make a distinction between engine and module experimental features.
Add additional telemetry points in load module.
Make CacheDirectory on Platform internal, we can decide to make it public later, if desired.

JamesWTruher committed:
Change filename for telemetry class

JamesWTruher committed:
add copyright to test file

PoshChan commented on build.psm1:
```diff
@@ -1012,7 +1012,7 @@ function Start-PSPester {
     # All concatenated commands/arguments are suffixed with the delimiter (space)
```
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-macos`
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @(@{InputObject=System.Private.Xml.Linq.dll; SideIndicator=<=}, @{InputObject=System.Threading.Timer.dll; SideIndicator=<=}, @{InputObject=System.Xml.XDocument.dll; SideIndicator=<=}).
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Host/Startup.Tests.ps1: line 123
123:         $diffs | Should -BeExactly $null
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-windows`
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @(@{InputObject=System.Private.Xml.Linq.dll; SideIndicator=<=}, @{InputObject=System.Threading.Timer.dll; SideIndicator=<=}, @{InputObject=System.Xml.XDocument.dll; SideIndicator=<=}).
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\Startup.Tests.ps1: line 123
123:         $diffs | Should -BeExactly $null
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-linux`
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @(@{InputObject=System.Private.Xml.Linq.dll; SideIndicator=<=}, @{InputObject=System.Threading.Timer.dll; SideIndicator=<=}, @{InputObject=System.Xml.XDocument.dll; SideIndicator=<=}).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/Startup.Tests.ps1: line 123
123:         $diffs | Should -BeExactly $null
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-macos`
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @(@{InputObject=System.Private.Xml.Linq.dll; SideIndicator=<=}, @{InputObject=System.Threading.Timer.dll; SideIndicator=<=}, @{InputObject=System.Xml.XDocument.dll; SideIndicator=<=}).
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Host/Startup.Tests.ps1: line 123
123:         $diffs | Should -BeExactly $null
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-windows`
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @(@{InputObject=System.Private.Xml.Linq.dll; SideIndicator=<=}, @{InputObject=System.Threading.Timer.dll; SideIndicator=<=}, @{InputObject=System.Xml.XDocument.dll; SideIndicator=<=}).
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\Startup.Tests.ps1: line 123
123:         $diffs | Should -BeExactly $null
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-linux`
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @(@{InputObject=System.Private.Xml.Linq.dll; SideIndicator=<=}, @{InputObject=System.Threading.Timer.dll; SideIndicator=<=}, @{InputObject=System.Xml.XDocument.dll; SideIndicator=<=}).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/Startup.Tests.ps1: line 123
123:         $diffs | Should -BeExactly $null
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-macos`
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @(@{InputObject=System.Private.Xml.Linq.dll; SideIndicator=<=}, @{InputObject=System.Threading.Timer.dll; SideIndicator=<=}, @{InputObject=System.Xml.XDocument.dll; SideIndicator=<=}).
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Host/Startup.Tests.ps1: line 123
123:         $diffs | Should -BeExactly $null
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-windows`
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @(@{InputObject=System.Private.Xml.Linq.dll; SideIndicator=<=}, @{InputObject=System.Threading.Timer.dll; SideIndicator=<=}, @{InputObject=System.Xml.XDocument.dll; SideIndicator=<=}).
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\Startup.Tests.ps1: line 123
123:         $diffs | Should -BeExactly $null
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-linux`
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @(@{InputObject=System.Private.Xml.Linq.dll; SideIndicator=<=}, @{InputObject=System.Threading.Timer.dll; SideIndicator=<=}, @{InputObject=System.Xml.XDocument.dll; SideIndicator=<=}).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/Startup.Tests.ps1: line 123
123:         $diffs | Should -BeExactly $null
```

PoshChan replied:
@JamesWTruher, your last commit had 2 failures in `PowerShell-CI-static-analysis`
Verify Markdown Links.Verify links in /home/vsts/work/1/s/docs/testing-guidelines/TestRoadmap.md.https://coveralls.io/github/OpenCover/opencover should work
```powershell
retry of URL failed with error: 
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 111
111:                                 throw "retry of URL failed with error: $($_.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/test/tools/CodeCoverageAutomation/README.md.https://coveralls.io/github/PowerShell/PowerShell should work
```powershell
retry of URL failed with error: 
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 111
111:                                 throw "retry of URL failed with error: $($_.Message)"

```

PoshChan replied:
@JamesWTruher, your last commit had 2 failures in `PowerShell-CI-static-analysis`
Verify Markdown Links.Verify links in /home/vsts/work/1/s/test/tools/CodeCoverageAutomation/README.md.https://coveralls.io/github/PowerShell/PowerShell should work
```powershell
retry of URL failed with error: 
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 111
111:                                 throw "retry of URL failed with error: $($_.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/docs/testing-guidelines/TestRoadmap.md.https://coveralls.io/github/OpenCover/opencover should work
```powershell
retry of URL failed with error: 
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 111
111:                                 throw "retry of URL failed with error: $($_.Message)"

```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-macos`
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @(@{InputObject=System.Private.Xml.Linq.dll; SideIndicator=<=}, @{InputObject=System.Threading.Timer.dll; SideIndicator=<=}, @{InputObject=System.Xml.XDocument.dll; SideIndicator=<=}).
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Host/Startup.Tests.ps1: line 123
123:         $diffs | Should -BeExactly $null
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-windows`
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @(@{InputObject=System.Private.Xml.Linq.dll; SideIndicator=<=}, @{InputObject=System.Threading.Timer.dll; SideIndicator=<=}, @{InputObject=System.Xml.XDocument.dll; SideIndicator=<=}).
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\Startup.Tests.ps1: line 123
123:         $diffs | Should -BeExactly $null
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-linux`
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @(@{InputObject=System.Private.Xml.Linq.dll; SideIndicator=<=}, @{InputObject=System.Threading.Timer.dll; SideIndicator=<=}, @{InputObject=System.Xml.XDocument.dll; SideIndicator=<=}).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/Startup.Tests.ps1: line 123
123:         $diffs | Should -BeExactly $null
```

PoshChan replied:
@JamesWTruher, your last commit had 2 failures in `PowerShell-CI-static-analysis`
Verify Markdown Links.Verify links in /home/vsts/work/1/s/test/tools/CodeCoverageAutomation/README.md.https://coveralls.io/github/PowerShell/PowerShell should work
```powershell
retry of URL failed with error: 
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 111
111:                                 throw "retry of URL failed with error: $($_.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/docs/testing-guidelines/TestRoadmap.md.https://coveralls.io/github/OpenCover/opencover should work
```powershell
retry of URL failed with error: 
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 111
111:                                 throw "retry of URL failed with error: $($_.Message)"

```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-macos`
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @{InputObject=.Xml.XDocument.dll; SideIndicator=<=}.
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Host/Startup.Tests.ps1: line 121
121:         $diffs | Should -BeExactly $null
```

PoshChan replied:
@JamesWTruher, your last commit had 2 failures in `PowerShell-CI-windows`
WindowStyle argument.-WindowStyle Maximized should work on Windows
```powershell
Expected exactly 'Maximized', but got Hidden.
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\ConsoleHost.Tests.ps1: line 839
839:             $showCmd | Should -BeExactly $WindowStyle
```
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @{InputObject=.Xml.XDocument.dll; SideIndicator=<=}.
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\Startup.Tests.ps1: line 121
121:         $diffs | Should -BeExactly $null
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-linux`
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @{InputObject=.Xml.XDocument.dll; SideIndicator=<=}.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/Startup.Tests.ps1: line 121
121:         $diffs | Should -BeExactly $null
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-macos`
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @{InputObject=.Xml.XDocument.dll; SideIndicator=<=}.
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Host/Startup.Tests.ps1: line 121
121:         $diffs | Should -BeExactly $null
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-windows`
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @{InputObject=.Xml.XDocument.dll; SideIndicator=<=}.
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\Startup.Tests.ps1: line 121
121:         $diffs | Should -BeExactly $null
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-linux`
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @{InputObject=.Xml.XDocument.dll; SideIndicator=<=}.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/Startup.Tests.ps1: line 121
121:         $diffs | Should -BeExactly $null
```

PoshChan replied:
@JamesWTruher, your last commit had 5 failures in `PowerShell-CI-windows`
Resources strings in Microsoft.PowerShell.ConsoleHost (was -ResGen used with Start-PSBuild).'ManagedEntranceStrings' should be an available type and the strings should be correct
```powershell
Expected strings to be the same, but they were different.
Expected length: 802
Actual length:   809
Strings differ at index 19.
Expected: '...xe] [-Logi...'
But was:  '...xe] [[-Fil...'
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\ResourceValidation\TestRunner.ps1: line 64
64:                     NormalizeLineEnd($resourceStringToCheck) | Should -Be (NormalizeLineEnd($inresource.value))
```
ConsoleHost unit tests.-Login pwsh switch.Accepts the -login switch for -Login and behaves correctly
```powershell
Expected $null or empty, but got @(<empty>, 'Usage: pwsh[.exe] [[-File] <filePath> [args]]', '                  [-Command { - | <script-block> [-args <arg-array>]', '                                | <string> [<CommandParameters>] } ]', '                  [-ConfigurationName <string>] [-CustomPipeName <string>]', '                  [-EncodedCommand <Base64EncodedCommand>]', '                  [-ExecutionPolicy <ExecutionPolicy>] [-InputFormat {Text | XML}]', '                  [-Interactive] [-LoadProfile] [-MTA] [-NoExit] [-NoLogo] [-NonInteractive] [-NoProfile]', '                  [-OutputFormat {Text | XML}]  [-SettingsFile <filePath>] [-STA] [-Version]', '                  [-WindowStyle <style>] [-WorkingDirectory <directoryPath>]').
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\ConsoleHost.Tests.ps1: line 302
302:                 $result | Should -BeNullOrEmpty
```
ConsoleHost unit tests.-Login pwsh switch.Accepts the -Login switch for -Login and behaves correctly
```powershell
Expected $null or empty, but got @(<empty>, 'Usage: pwsh[.exe] [[-File] <filePath> [args]]', '                  [-Command { - | <script-block> [-args <arg-array>]', '                                | <string> [<CommandParameters>] } ]', '                  [-ConfigurationName <string>] [-CustomPipeName <string>]', '                  [-EncodedCommand <Base64EncodedCommand>]', '                  [-ExecutionPolicy <ExecutionPolicy>] [-InputFormat {Text | XML}]', '                  [-Interactive] [-LoadProfile] [-MTA] [-NoExit] [-NoLogo] [-NonInteractive] [-NoProfile]', '                  [-OutputFormat {Text | XML}]  [-SettingsFile <filePath>] [-STA] [-Version]', '                  [-WindowStyle <style>] [-WorkingDirectory <directoryPath>]').
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\ConsoleHost.Tests.ps1: line 302
302:                 $result | Should -BeNullOrEmpty
```
ConsoleHost unit tests.-Login pwsh switch.Accepts the -LOGIN switch for -Login and behaves correctly
```powershell
Expected $null or empty, but got @(<empty>, 'Usage: pwsh[.exe] [[-File] <filePath> [args]]', '                  [-Command { - | <script-block> [-args <arg-array>]', '                                | <string> [<CommandParameters>] } ]', '                  [-ConfigurationName <string>] [-CustomPipeName <string>]', '                  [-EncodedCommand <Base64EncodedCommand>]', '                  [-ExecutionPolicy <ExecutionPolicy>] [-InputFormat {Text | XML}]', '                  [-Interactive] [-LoadProfile] [-MTA] [-NoExit] [-NoLogo] [-NonInteractive] [-NoProfile]', '                  [-OutputFormat {Text | XML}]  [-SettingsFile <filePath>] [-STA] [-Version]', '                  [-WindowStyle <style>] [-WorkingDirectory <directoryPath>]').
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\ConsoleHost.Tests.ps1: line 302
302:                 $result | Should -BeNullOrEmpty
```
ConsoleHost unit tests.-Login pwsh switch.Accepts the -log switch for -Login and behaves correctly
```powershell
Expected 0, but got 64.
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\ConsoleHost.Tests.ps1: line 303
303:                 $LASTEXITCODE | Should -Be 0
```

iSazonov commented:
> Update startup telemetry to measure the type of shell startup (remote/ssh)

We will have `-Login` in days. That might be helpful too.

JamesWTruher committed:
Partial updates for Codacy and Code Factor

JamesWTruher commented on build.psm1:
```diff
@@ -1012,7 +1012,7 @@ function Start-PSPester {
 
     # Disable telemetry for all startups of pwsh in tests
-    $command = "`$env:POWERSHELL_TELEMETRY_OPTOUT = 1;"
+    $command = "`$env:POWERSHELL_TELEMETRY_OPTOUT = 'yes';"
```
just to make it more clear

JamesWTruher committed:
Remove commented reference to application insights package

JamesWTruher commented on src/Microsoft.PowerShell.ConsoleHost/Microsoft.PowerShell.ConsoleHost.csproj:
```diff
@@ -9,7 +9,7 @@
     <ProjectReference Include="..\System.Management.Automation\System.Management.Automation.csproj" />
     <!-- the following package(s) are from https://github.com/microsoft/applicationinsights-??? -->
-    <PackageReference Include="Microsoft.ApplicationInsights" Version="2.10.0" />
+    <!--- <PackageReference Include="Microsoft.ApplicationInsights" Version="2.10.0" /> -->
```
fixed

JamesWTruher committed:
More CodeFactor fixes

adityapatwardhan commented on test/powershell/engine/Basic/Telemetry.Tests.ps1:
```diff
@@ -0,0 +1,137 @@
+    It "Should properly set whether telemetry is sent based on when environment variable is not set" -TestCases $telemetryIsSetData {
+        param ( [string]$name, [string]$value, [string]$expectedValue )
+        if ( Test-Path -Path env:POWERSHELL_TELEMETRY_OPTOUT ) {
+            Remote-Item -Path env:POWERSHELL_TELEMETRY_OPTOUT
```
Is this needed, `AfterEach` seems to be doing this.

JamesWTruher replied:
actually, this test had a more serious problem as telemetryIsSetData wasn't actually assigned when we got to this test. I've rewritten it

JamesWTruher committed:
more code factor fixes

JamesWTruher committed:
Update test to reflect unloaded assemblies

JamesWTruher committed:
Update test/powershell/engine/Basic/Telemetry.Tests.ps1

Co-Authored-By: Aditya Patwardhan <adityap@microsoft.com>

JamesWTruher committed:
Update src/System.Management.Automation/engine/GetCommandCommand.cs

Co-Authored-By: Travis Plunk <travis.plunk@microsoft.com>

JamesWTruher commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -0,0 +1,361 @@
+            properties.Add("UUID", s_uniqueUserIdentifier.ToString());
+            properties.Add("GitCommitID", PSVersionInfo.GitCommitId);
+            properties.Add("OSDescription", RuntimeInformation.OSDescription);
+            properties.Add("OSChannel", (channel == null || channel == string.Empty) ? "unknown" : channel);
```
yah - that's better

JamesWTruher commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -0,0 +1,361 @@
+            properties.Add("GitCommitID", PSVersionInfo.GitCommitId);
+            properties.Add("OSDescription", RuntimeInformation.OSDescription);
+            properties.Add("OSChannel", (channel == null || channel == string.Empty) ? "unknown" : channel);
+            properties.Add("StartMode", (mode == null || mode == string.Empty) ? "unknown" : mode);
```
fixed

JamesWTruher committed:
additional codefactor issues

JamesWTruher committed:
Remove assemblies which are no longer loaded

JamesWTruher committed:
Harden link tests by allowing 503 response codes

JamesWTruher committed:
fix typo with regard to System.Xml.XDocument.dll

the entire assembly should be removed

JamesWTruher committed:
make anonymous a const string

JamesWTruher commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -0,0 +1,361 @@
+                return moduleNameToValidate;
+            }
+            else {
+                return "anonymous";
```
fixed

JamesWTruher commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -0,0 +1,361 @@
+                    return featureNameToValidate;
+                }
+            }
+            return "anonymous";
```
fixed

JamesWTruher commented on src/System.Management.Automation/engine/hostifaces/ConnectionBase.cs:
```diff
@@ -262,6 +262,7 @@ private void CoreOpen(bool syncCall)
             OpenHelper(syncCall);
             if (etwEnabled) RunspaceEventSource.Log.OpenRunspaceStop();
 
+            // Telemetry here - runspace start
```
we decided to do it in PowerShell.Create instead rather than catching the runspace side.

JamesWTruher replied:
removed

JamesWTruher committed:
Fixes to tests to address feedback

Use Should -Exist rather than test-path
Fix test when environment variable is not set

JamesWTruher commented on test/powershell/engine/Basic/Telemetry.Tests.ps1:
```diff
@@ -0,0 +1,137 @@
+    It "Should create a uuid file if telemetry is opted in" {
+        $env:POWERSHELL_TELEMETRY_OPTOUT = "no"
+        & $PWSH -command "exit"
+        Test-Path -Path $uuidPath  | Should -Be $true
```
fixed

JamesWTruher commented on test/powershell/engine/Basic/Telemetry.Tests.ps1:
```diff
@@ -0,0 +1,137 @@
+    It "Should create a property uuid file when telemetry is sent" {
+        $env:POWERSHELL_TELEMETRY_OPTOUT = "no"
+        & $PWSH -command "exit"
+        Test-Path -Path $uuidPath  | Should -Be $true
```
fixed

iSazonov commented on test/powershell/engine/Basic/Telemetry.Tests.ps1:
```diff
@@ -0,0 +1,132 @@
+
+    It "Should not create a uuid file if telemetry is opted out" {
+        $env:POWERSHELL_TELEMETRY_OPTOUT = 1
+        & $PWSH -command "exit"
```
```suggestion
        & $PWSH -NoProfile -Command "exit"
```

iSazonov commented on test/powershell/engine/Basic/Telemetry.Tests.ps1:
```diff
@@ -0,0 +1,132 @@
+
+    It "Should create a uuid file if telemetry is opted in" {
+        $env:POWERSHELL_TELEMETRY_OPTOUT = "no"
+        & $PWSH -command "exit"
```
```suggestion
        & $PWSH -NoProfile -Command "exit"
```

iSazonov commented on test/powershell/engine/Basic/Telemetry.Tests.ps1:
```diff
@@ -0,0 +1,132 @@
+
+    It "Should create a uuid file by default" {
+        if ( Test-Path env:POWERSHELL_TELEMETRY_OPTOUT ) { Remove-Item -Path env:POWERSHELL_TELEMETRY_OPTOUT }
+        & $PWSH -command "exit"
```
```suggestion
        & $PWSH -NoProfile -Command "exit"
```

iSazonov commented on test/powershell/engine/Basic/Telemetry.Tests.ps1:
```diff
@@ -0,0 +1,132 @@
+
+    It "Should create a property uuid file when telemetry is sent" {
+        $env:POWERSHELL_TELEMETRY_OPTOUT = "no"
+        & $PWSH -command "exit"
```
```suggestion
        & $PWSH -NoProfile -Command "exit"
```

iSazonov commented on test/powershell/engine/Basic/Telemetry.Tests.ps1:
```diff
@@ -0,0 +1,132 @@
+    It "Should not create a telemetry file if one already exists and telemetry is opted in" {
+        [byte[]]$bytes = [System.Guid]::NewGuid().ToByteArray()
+        [System.IO.File]::WriteAllBytes($uuidPath, $bytes)
+        & $PWSH -command "exit"
```
```suggestion
        & $PWSH -NoProfile -Command "exit"
```

iSazonov commented on test/powershell/engine/Basic/Telemetry.Tests.ps1:
```diff
@@ -0,0 +1,132 @@
+    It "Should create a new telemetry file if the current one is 00000000-0000-0000-0000-000000000000" {
+        [byte[]]$zeroGuid = [System.Guid]::Empty.ToByteArray()
+        [System.IO.File]::WriteAllBytes($uuidPath, $zeroGuid)
+        & $PWSH -command "exit"
```
```suggestion
        & $PWSH -NoProfile -Command "exit"
```

iSazonov commented on test/powershell/engine/Basic/Telemetry.Tests.ps1:
```diff
@@ -0,0 +1,132 @@
+    It "Should create a new telemetry file if the current one is smaller than 16 bytes" {
+        $badBytes = [byte[]]::new(8);
+        [System.IO.File]::WriteAllBytes($uuidPath, $badBytes)
+        & $PWSH -command "exit"
```
```suggestion
        & $PWSH -NoProfile -Command "exit"
```

iSazonov commented on test/powershell/engine/Basic/Telemetry.Tests.ps1:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    It "Should properly set whether telemetry is sent based on when environment variable is not set" {
+        $result = & $PWSH -c '[Microsoft.PowerShell.Telemetry.ApplicationInsightsTelemetry]::CanSendTelemetry'
```
```suggestion
        $result = & $PWSH -NoProfile -Command [Microsoft.PowerShell.Telemetry.ApplicationInsightsTelemetry]::CanSendTelemetry'
```

iSazonov commented on test/powershell/engine/Basic/Telemetry.Tests.ps1:
```diff
@@ -0,0 +1,132 @@
+    It "Should properly set whether telemetry is sent based on environment variable when <name>" -TestCases $telemetryIsSetData {
+        param ( [string]$name, [string]$value, [string]$expectedValue )
+        $env:POWERSHELL_TELEMETRY_OPTOUT = $value
+        $result = & $PWSH -c '[Microsoft.PowerShell.Telemetry.ApplicationInsightsTelemetry]::CanSendTelemetry'
```
```suggestion
        $result = & $PWSH -NoProfile -Command '[Microsoft.PowerShell.Telemetry.ApplicationInsightsTelemetry]::CanSendTelemetry'
```

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/Telemetry.cs:
```diff
@@ -1,102 +0,0 @@
-        static ApplicationInsightsTelemetry()
-        {
-            TelemetryConfiguration.Active.InstrumentationKey = _psCoreTelemetryKey;
-            TelemetryConfiguration.Active.TelemetryChannel.DeveloperMode = _developerMode;
```
@JamesWTruher It turns out this two lines caused the following 3 assemblies to be loaded at the startup time.
In the new changes, you avoid running these 2 lines of code when the telemetry is turned off (_CI builds always turn off telemetry_), and hence the following 3 assemblies are not loaded at startup now.

This explains the failure to that startup test, but also reveals a problem -- that test is not tracking the real startup scenario, but only the telemetry-optout scenario.
Today, with the telemetry turned on, the telemetry code results in 6 assemblies being loaded at startup time.
I think we should run `startup.test.ps1` with telemetry turned on, but that will produce some noise to our telemetry data.

```
System.Private.Xml.Linq.dll
System.Threading.Timer.dll
System.Xml.XDocument.dll
```

```
Microsoft.ApplicationInsight.dll
System.Diagnostics.DiagnosticSource.dll
System.Net.NameResolution.dll
```

JamesWTruher commented on test/powershell/engine/Basic/Telemetry.Tests.ps1:
```diff
@@ -0,0 +1,137 @@
+    It "Should not create a uuid file if telemetry is opted out" {
+        $env:POWERSHELL_TELEMETRY_OPTOUT = 1
+        & $PWSH -command "exit"
+        Test-Path -Path $uuidPath  | Should -Be $false
```
fixed

JamesWTruher committed:
address feedback to call powershell without profile '-noprofile'

fix case of '-Command'
address other codefactor issue

JamesWTruher commented:
@adityapatwardhan , @TravisEz13  I believe that I have addressed your blocking issues.

adityapatwardhan commented:
@JamesWTruher Please have a look at failing windows tests

JamesWTruher committed:
Turn off developer mode at the last
fix one more test to exclude profile

JamesWTruher committed:
Remove comment about telemetry collection in this location

We put it in the PowerShell.Create code rather than in this location

iSazonov commented on test/powershell/engine/Basic/Telemetry.Tests.ps1:
```diff
@@ -0,0 +1,132 @@
+        $result = & $PWSH -NoProfile -Command '[Microsoft.PowerShell.Telemetry.ApplicationInsightsTelemetry]::CanSendTelemetry'
+        $result | Should -Be $expectedValue
+    }
+
```
Extra line
```suggestion
```

iSazonov replied:
Extra line
```suggestion
```

iSazonov reviewed (approved):
LGTM with one minor comment.

iSazonov commented:
#10050 was merged and now it is possible to add telemetry in the code too.

JamesWTruher committed:
Update test/powershell/engine/Basic/Telemetry.Tests.ps1

Co-Authored-By: Ilya <darpa@yandex.ru>

adityapatwardhan commented:
@JamesWTruher All looks good and ready to merge. As @iSazonov said above, are the plans to add telemetry to `pwsh -login`?

JamesWTruher commented:
@iSazonov @adityapatwardhan We can add additional telemetry for login to a subsequent PR if we decide it is worth knowing. I don't believe that we must do that now, though.

JamesWTruher committed:
Merge branch 'master' into telemetry20

daxian-dbw commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -0,0 +1,407 @@
+
+            string metricName = metricId.ToString();
+            string uuidString = s_uniqueUserIdentifier.ToString();
+            string sessionIdString  = s_sessionId.ToString();
```
I recommend to make `s_uniqueUserIdentifier` and `s_sessionId` string type, so we only call `ToString()` on the GUID once. Now we are creating new strings every time `SendTelemetryMetric` is called.

JamesWTruher replied:
fixed

daxian-dbw commented on src/System.Management.Automation/engine/GetCommandCommand.cs:
```diff
@@ -554,6 +554,8 @@ private void OutputResultsHelper(IEnumerable<CommandInfo> results)
 
             _timer.Stop();
```
Since the tracked elapsed time is not used at all, shall we remove the `_timer` and the related time tracking code?

JamesWTruher replied:
fixed

daxian-dbw commented on src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs:
```diff
@@ -591,6 +591,8 @@ private static CommandCompletion CompleteInputImpl(Ast ast, Token[] tokens, IScr
 
                     var completionResults = results ?? EmptyCompletionResult;
                     sw.Stop();
```
Same here, shall we remove the timer code?

JamesWTruher replied:
fixed

daxian-dbw commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -0,0 +1,407 @@
+                // use a hashset when looking for module names, it should be quicker than a string comparison
+                s_knownModules = new HashSet<string>(s_knownModuleNames, StringComparer.OrdinalIgnoreCase);
+                s_uniqueUserIdentifier = GetUniqueIdentifier();
+                s_developerMode = false;
```
Do we need this static field? It's not used anywhere else, so maybe just remove it and change to `TelemetryConfiguration.Active.TelemetryChannel.DeveloperMode = false`?

JamesWTruher replied:
fixed

JamesWTruher committed:
Address feedback from @daxian-dbw

I don't need to keep the static strings of allowed module names or do convoluted checking of the experimental feature name.
Move the initialization of the HashSet with list of module names into the static constructor.
Ifdef a couple stopwatches out of the code (keeping the location tagged if we want to bring that back).
Add a comment about POWERSHELL_DISTRIBUTION_CHANNEL.
Change the type of the sessionid and uniqueid into strings so I don't repeatedly call ToString.

JamesWTruher commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -0,0 +1,407 @@
+        // Get the module name. If we can report it, we'll return the name, otherwise, we'll return "anonymous"
+        private static string GetModuleName(string moduleNameToValidate)
+        {
+            if (s_knownModules.TryGetValue(moduleNameToValidate, out string valOut))
```
fixed

JamesWTruher commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -0,0 +1,407 @@
+                // An experimental feature in a module is guaranteed to start with the module name
+                // but we can't look up the experimental feature in the hash because there will be more than just the module name.
+                // we can't use a hashset because it's a partial comparison.
+                if (featureNameToValidate.StartsWith(knownModuleName, StringComparison.OrdinalIgnoreCase))
```
fixed

JamesWTruher commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -0,0 +1,407 @@
+                    case TelemetryType.PowerShellCreate:
+                    case TelemetryType.RemoteSessionOpen:
+                    case TelemetryType.ExperimentalEngineFeatureActivation:
+                        s_telemetryClient.GetMetric(metricName, "uuid", "SessionId", "Detail").TrackValue(1, uuidString, sessionIdString, data);
```
fixed

JamesWTruher commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -0,0 +1,407 @@
+                return;
+            }
+            var properties = new Dictionary<string, string>();
+            var channel = Environment.GetEnvironmentVariable("POWERSHELL_DISTRIBUTION_CHANNEL");
```
fixed

daxian-dbw commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -0,0 +1,403 @@
+                        "ThreadJob",
+                    },
+                    StringComparer.OrdinalIgnoreCase
+                    );
```
I recommend to use the collection initializer syntax, as in
```c#
s_knownModules = new HashSet<string>(StringComparer.OrdinalIgnoreCase) {
    "Microsoft.PowerShell.Archive",
    "Microsoft.PowerShell.Host",
    ...
};
```

JamesWTruher replied:
fixed

daxian-dbw commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -0,0 +1,403 @@
+            }
+
+            string metricName = metricId.ToString();
+            string uuidString = s_uniqueUserIdentifier;
```
It seems we can use `s_uniqueUserIdentifier` in the `TrackValue` calls directly without needing `uuidString`.

JamesWTruher replied:
fixed

daxian-dbw commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -0,0 +1,403 @@
+            // An experimental feature in a module is guaranteed to start with the module name
+            // we can strip out the text past the last '.' as the text before that will be the ModuleName
+            int lastDotIndex = featureNameToValidate.LastIndexOf('.');
+            string moduleName = featureNameToValidate.Substring(0,lastDotIndex);
```
> 0,lastDotIndex

Can you please add a space after the comma? CodeFacotr might complain :)

JamesWTruher replied:
fixed

daxian-dbw commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -0,0 +1,403 @@
+            else
+            {
+                return _anonymous;
+            }
```
Minor comment: we can remove the `else` block and just have `return _anonymous`.

JamesWTruher replied:
yes!

JamesWTruher committed:
Don't create another string, just use s_uniqueUserIdentifier directly.
Optimize hashset creation to not create a new string array.
